math/mathcore/inc/Math/CladDerivator.h

```diff
@@ -19,420 +19,215 @@
-    return 1./ TMath::Sq(TMath::Cos(d));
-  }
-
-  template <typename T>
-  Double_t Tan_darg0_darg0(T d) {
-    return 2. * TMath::Sin(d) / (TMath::Cos(d) * TMath::Cos(d) * TMath::Cos(d));
-  }
-
-  template <typename T>
-  Double_t TanH_darg0(T d) {
-    return 1./ TMath::Sq(TMath::CosH(d));
-  }
-
-  template <typename T>
-  Double_t TanH_darg0_darg0(T d) {
-    return -2. * TMath::SinH(d) / (TMath::CosH(d) * TMath::CosH(d) * TMath::CosH(d));
-  }
+namespace TMath {
+template <typename T>
+ValueAndPushforward<T, T> Abs_pushforward(T x, T d_x)
+{
+   return {::TMath::Abs(x), ((x < 0) ? -1 : 1) * d_x};
+}
+
+template <typename T>
+ValueAndPushforward<T, T> ACos_pushforward(T x, T d_x)
+{
+   return {::TMath::ACos(x), (-1. / ::TMath::Sqrt(1 - x * x)) * d_x};
+}
+
+template <typename T>
+ValueAndPushforward<T, T> ACosH_pushforward(T x, T d_x)
+{
+   return {::TMath::ACosH(x), (1. / ::TMath::Sqrt(x * x - 1)) * d_x};
+}
+
+template <typename T>
+ValueAndPushforward<T, T> ASin_pushforward(T x, T d_x)
+{
+   return {::TMath::ASin(x), (1. / ::TMath::Sqrt(1 - x * x)) * d_x};
+}
+
+template <typename T>
+ValueAndPushforward<T, T> ASinH_pushforward(T x, T d_x)
+{
+   return {::TMath::ASinH(x), (1. / ::TMath::Sqrt(x * x + 1)) * d_x};
+}
+
+template <typename T>
+ValueAndPushforward<T, T> ATan_pushforward(T x, T d_x)
+{
+   return {::TMath::ATan(x), (1. / (x * x + 1)) * d_x};
+}
+
+template <typename T>
+ValueAndPushforward<T, T> ATanH_pushforward(T x, T d_x)
+{
+   return {::TMath::ATanH(x), (1. / (1 - x * x)) * d_x};
+}
+
+template <typename T>
+ValueAndPushforward<T, T> Cos_pushforward(T x, T d_x)
+{
+   return {::TMath::Cos(x), -::TMath::Sin(x) * d_x};
 }
+
+template <typename T>
+ValueAndPushforward<T, T> CosH_pushforward(T x, T d_x)
+{
+   return {::TMath::CosH(x), ::TMath::SinH(x) * d_x};
+}
+
+template <typename T>
+ValueAndPushforward<T, T> Erf_pushforward(T x, T d_x)
+{
+   return {::TMath::Erf(x), (2 * ::TMath::Exp(-x * x) / ::TMath::Sqrt(::TMath::Pi())) * d_x};
+}
+
+template <typename T>
+ValueAndPushforward<T, T> Erfc_pushforward(T x, T d_x)
+{
+   return {::TMath::Erfc(x), -Erf_pushforward(x, d_x).pushforward};
+}
+
+template <typename T>
+ValueAndPushforward<T, T> Exp_pushforward(T x, T d_x)
+{
+   return {::TMath::Exp(x), ::TMath::Exp(x) * d_x};
+}
+
+template <typename T>
+ValueAndPushforward<T, T> Hypot_pushforward(T x, T y, T d_x, T d_y)
+{
+   return {::TMath::Hypot(x, y), x / ::TMath::Hypot(x, y) * d_x + y / ::TMath::Hypot(x, y) * d_y};
+}
+
+template <typename T, typename U>
+void Hypot_pullback(T x, T y, U p, clad::array_ref<T> d_x, clad::array_ref<T> d_y)
+{
+   T h = ::TMath::Hypot(x, y);
+   *d_x += x / h * p;
+   *d_y += y / h * p;
+}
+
+template <typename T>
+ValueAndPushforward<T, T> Log_pushforward(T x, T d_x)
+{
+   return {::TMath::Log(x), (1. / x) * d_x};
+}
+
+template <typename T>
+ValueAndPushforward<T, T> Log10_pushforward(T x, T d_x)
+{
+   return {::TMath::Log10(x), (1.0 / (x * ::TMath::Ln10())) * d_x};
```

Comment:
Isn't the derivative of `Ln10` 0 and thus we divide by 0 for the second order derivative?

Reply:
We are derivating log<sub>10</sub>(x).
The result of first order derivative is ```1.0/ x* log(10)```. We are not dividing by zero unless x is 0.